test/unit/meson.build

```
@@ -0,0 +1,24 @@
+cmocka_deps = dependency('cmocka', version : '>=1.0.0')
```

```suggestion
cmocka_deps = dependency('cmocka', version : '>=1.0.0', default_options: ['default_library=static'])
```

And run `meson wrap install cmocka` to acquire the right file to make sure that Meson >= 0.55.0 can automatically fall back to downloading, checksumming, and building a private copy if it doesn't find a system one.

The default options ensures that when building as a subproject fallback it doesn't rather pointlessly build a shared library for cmocka.

---

Thanks for the suggestion, and great to see you here!

I tried that. It definitely works on darwin if I uninstall the cmocka brew package and remove the libui-ng fat binary hack.

Sadly is still fails with mingw.
If I leave the mingw-w64-cmocka package intact it will pull in the dynamic library.
If i uninstall it, it actually tries to pull in the linux headers in my case, but that might be a bug in `x86_64-w64-mingw32-meson` though.

---

> Thanks for the suggestion, and great to see you here!

I'm always happy to give people Meson tips. :)

> Sadly is still fails with mingw.

You can select to use the subproject even if mingw-w64-cmocka is installed. Meson provides fine-grained control for selecting between system and fallback versions:

```
# to force using fallbacks whenever possible, good for distributing standalone binaries that bundle dependencies
meson setup builddir/ --wrap-mode=forcefallback

# to force one specific fallback to be used
meson setup builddir/ --force-fallback-for=cmocka
```

I don't know what the linux headers issue is there. Logs might help I suppose -- because I don't have a Windows box, so I can't try running it to create those logs myself. :(

---

> ```
> # to force using fallbacks whenever possible, good for distributing standalone binaries that bundle dependencies
> meson setup builddir/ --wrap-mode=forcefallback
> 
> # to force one specific fallback to be used
> meson setup builddir/ --force-fallback-for=cmocka
> ```

Wow, yeah that actually works. Sweet. Still a lot of magic going on, but it looks like we can lock dependencies down and verify by hash.

> I don't know what the linux headers issue is there. Logs might help I suppose -- because I don't have a Windows box, so I can't try running it to create those logs myself. :(

This is actually from an ArchLinux box. mingw-w64 cross compile. Somehow the `mingw-w64-meson` package will use the systems `cmocka` library if it does not find a mingw  (`mingw-w64`) one. Which results in `-I/usr/include` erroneously getting passed to the compiler, causing havoc. Cmocka does not actually provide a proper way of building a static library and  `mingw-w64-cmocka` gets it wrong and produces a `cmocka-static.a`, which meson then can't find. I am not sure yet where to file a bug report yet. Probably in all projects.

---

> This is actually from an ArchLinux box. mingw-w64 cross compile. Somehow the `mingw-w64-meson` package will use the systems `cmocka` library if it does not find a mingw (`mingw-w64`) one. Which results in `-I/usr/include` erroneously getting passed to the compiler, causing havoc. Cmocka does not actually provide a proper way of building a static library and `mingw-w64-cmocka` gets it wrong and produces a `cmocka-static.a`, which meson then can't find. I am not sure yet where to file a bug report yet. Probably in all projects.

Oh yeah, that makes sense. :D I overlooked your above mention of Arch Linux and somehow thought you were using the MSYS2 shell. That package manager also has a mingw-w64-cmocka package...

So I *can*, as it happens, emulate your experience. However, what I discovered is as follows:
- if it is not installed, Meson does NOT pick up a system version via `/usr/bin/x86_64-w64-mingw32-pkg-config` (it does not detect my x86_64 linux version)
- if it *is* installed, it picks up the mingw sysroot version
  - but only the shared DLL. Unfortunately, the mingw sysroot isn't set up OOTB, so in this case you need to add `/usr/x86_64-w64-mingw32/bin` to $WINEPATH in order for cmocka.dll to be available when running the test program. If you do this, everything works.
  - the static lib is basically broken as you observed. I do not understand why upstream cmocka generates this broken name, it's totally wrong, but it's absolutely not the AUR package's fault, because upstream cmocka has an option -DBUILD_STATIC_LIB=ON that does: `add_library(cmocka-static STATIC ${cmocka_SRCS})`

---

I created issue in MSYS2: https://github.com/msys2/MINGW-packages/issues/11912